ext/mbstring/libmbfl/filters/mbfilter_utf16.c

```diff
@@ -322,3 +333,182 @@ static int mbfl_filt_conv_utf16_wchar_flush(mbfl_convert_filter *filter)
 
 	return 0;
 }
+
+static size_t mb_utf16_to_wchar(unsigned char **in, size_t *in_len, uint32_t *out, size_t out_len, unsigned int *state)
+{
+	if (*in_len >= 2) {
+		unsigned char *p = *in;
+		unsigned char c1 = *p++;
+		unsigned char c2 = *p++;
+		uint16_t n = (c1 << 8) | c2;
+
+		if (n == 0xFFFE) {
+			/* Little-endian BOM */
+			*in = p;
+			*in_len -= 2;
+			return mb_utf16le_to_wchar(in, in_len, out, out_len, NULL);
+		} if (n == 0xFEFF) {
+			/* Big-endian BOM; don't send to output */
+			*in = p;
+			*in_len -= 2;
+		}
+	}
+
+	return mb_utf16be_to_wchar(in, in_len, out, out_len, NULL);
+}
+
+static size_t mb_utf16be_to_wchar(unsigned char **in, size_t *in_len, uint32_t *buf, size_t bufsize, unsigned int *state)
+{
+	/* We only want to read 16-bit words out of `str`; any trailing byte will be handled at the end */
+	unsigned char *p = *in, *e = p + (*in_len & ~1);
+	/* Set `limit` to one less than the actual amount of space in the buffer; this is because
+	 * on some iterations of the below loop, we might produce two output words */
+	uint32_t *out = buf, *limit = buf + bufsize - 1;
+
+	while (p < e && out < limit) {
+		unsigned char c1 = *p++;
+		unsigned char c2 = *p++;
+		uint16_t n = (c1 << 8) | c2;
+
+		if (n >= 0xD800 && n <= 0xDBFF) {
+			/* Handle surrogate */
+			if (p < e) {
+				unsigned char c3 = *p++;
+				unsigned char c4 = *p++;
+				uint16_t n2 = (c3 << 8) | c4;
+
+				if (n2 >= 0xD800 && n2 <= 0xDBFF) {
+					/* Wrong; that's the first half of a surrogate pair, when we were expecting the second */
+					*out++ = MBFL_BAD_INPUT;
+					p -= 2;
+				} else if (n2 >= 0xDC00 && n2 <= 0xDFFF) {
+					*out++ = (((n & 0x3FF) << 10) | (n2 & 0x3FF)) + 0x10000;
+				} else {
+					/* The first half of a surrogate pair was followed by a 'normal' codepoint */
+					*out++ = MBFL_BAD_INPUT;
+					*out++ = n2;
+				}
+			} else {
+				*out++ = MBFL_BAD_INPUT;
+			}
+		} else if (n >= 0xDC00 && n <= 0xDFFF) {
+			/* This is wrong; second part of surrogate pair has come first */
+			*out++ = MBFL_BAD_INPUT;
+		} else {
+			*out++ = n;
+		}
+	}
+
+	if (p == e && (*in_len & 0x1) && out < limit) {
+		/* There is an extra trailing byte (which shouldn't be there) */
+		*out++ = MBFL_BAD_INPUT;
+		p++;
+	}
```

Comment:
Something to note here is that the API as currently designed doesn't allow for streaming use, i.e. where the input string is not provided in its entirety upfront. You can't feed only a partial string into to_wchar, because it would be interpreted as the string end. We'd need an `end` flag to support this.

Are we okay with that? Do we have any streaming uses of the old API?

Reply:
That was deliberate, but if anyone knows of cases where 'streaming' use is needed, this is the time to speak up!

I am not aware of any 'streaming' use of the old API. It's always used on complete strings in memory.